src/script.rs

```diff
@@ -265,219 +260,113 @@ impl From<Operation> for u8 {
+                return Ok(i64::MIN);
+            };
 
-impl From<u32> for ScriptNum {
-    fn from(value: u32) -> Self {
-        ScriptNum(value.into())
-    }
-}
+            // Ensure defined behaviour (in Rust, left shift of `i64` by 64 bits is an arithmetic
+            // overflow that may panic or give an unspecified result). The above encoding of
+            // `i64::MIN` is the only allowed 9-byte encoding.
+            if vch.len() > 8 {
+                return Err(ScriptNumError::Overflow {
+                    max_num_size: 8,
+                    actual: vch.len(),
+                });
+            };
 
-impl From<u8> for ScriptNum {
-    fn from(value: u8) -> Self {
-        ScriptNum(value.into())
-    }
-}
+            let mut result: i64 = 0;
+            for (i, vch_i) in vch.iter().enumerate() {
+                result |= i64::from(*vch_i) << (8 * i);
+            }
 
-/// TODO: This instance will be obsolete if we convert bool directly to a `Vec<u8>`, which is also
-///       more efficient.
-impl From<bool> for ScriptNum {
-    fn from(value: bool) -> Self {
-        ScriptNum(value.into())
-    }
-}
+            // If the input vector's most significant byte is 0x80, remove it from the result's msb
+            // and return a negative.
+            if vch_back & 0x80 != 0 {
+                return Ok(-(result & !(0x80 << (8 * (vch.len() - 1)))));
+            };
 
-impl TryFrom<usize> for ScriptNum {
-    type Error = TryFromIntError;
-    fn try_from(value: usize) -> Result<Self, Self::Error> {
-        value.try_into().map(ScriptNum)
+            Ok(result)
+        }
     }
 }
 
-impl TryFrom<ScriptNum> for u16 {
-    type Error = TryFromIntError;
-    fn try_from(value: ScriptNum) -> Result<Self, Self::Error> {
-        value.getint().try_into()
+pub fn serialize_num(value: i64) -> Vec<u8> {
+    if value == 0 {
+        return Vec::new();
     }
-}
 
-impl TryFrom<ScriptNum> for u8 {
-    type Error = TryFromIntError;
-    fn try_from(value: ScriptNum) -> Result<Self, Self::Error> {
-        value.getint().try_into()
+    if value == i64::MIN {
+        // The code below was based on buggy C++ code, that produced the "wrong" result for
+        // INT64_MIN. In that case we intentionally return the result that the C++ code as compiled
+        // for zcashd (with `-fwrapv`) originally produced on an x86_64 system.
+        return vec![0, 0, 0, 0, 0, 0, 0, 128, 128];
     }
-}
 
-impl Add for ScriptNum {
-    type Output = Self;
-
-    fn add(self, other: Self) -> Self {
-        Self(
-            self.0
-                .checked_add(other.0)
-                .expect("caller should avoid overflow"),
-        )
+    let mut result = Vec::new();
+    let neg = value < 0;
+    let mut absvalue = value.abs();
+
+    while absvalue != 0 {
+        result.push(
+            (absvalue & 0xff)
+                .try_into()
+                .unwrap_or_else(|_| unreachable!()),
+        );
```

Comment:
I agree with @nuttycom. In this case it helps to see that the out-of-range cannot occur. Also prefer `expect` to `.unwrap_or_else(|_| unreachable!())` because the message to "expect" can help to clarify why there cannot be a panic.

```suggestion
        result.push(u8::try_from(absvalue & 0xff).expect("fits in u8"));
```

Reply:
Resolving as a nit so that this can merge.